src/cart/crt_iv.c

```diff
@@ -2993,6 +3025,133 @@ bulk_update_transfer_done(const struct crt_bulk_cb_info *info)
 	return rc;
 }
 
+static void
+crt_iv_update_inline_hdlr(void *arg)
+{
+	crt_rpc_t			*rpc = arg;
+	struct crt_ivns_internal	*ivns_internal;
+	struct crt_iv_ops		*iv_ops;
+	struct crt_iv_update_in		*input;
+	struct crt_iv_update_out	*output;
+	struct crt_ivns_id		ivns_id;
+	d_sg_list_t			iv_value = {0};
+	struct update_cb_info		*update_cb_info = NULL;
+	int				rc = 0;
+	d_rank_t			next_rank;
+	crt_iv_sync_t			*sync_type;
+	void				*user_priv;
+	uint32_t			grp_ver;
+
+	input = crt_req_get(rpc);
+	output = crt_reply_get(rpc);
+
+	D_ASSERT(input != NULL);
+	D_ASSERT(output != NULL);
+
+	ivns_id.ii_group_name = input->ivu_ivns_group;
+	ivns_id.ii_nsid = input->ivu_ivns_id;
+
+	/* ADDREF */
+	ivns_internal = crt_ivns_internal_lookup(&ivns_id);
+	D_ASSERT(ivns_internal != NULL);
+
+	iv_ops = crt_iv_ops_get(ivns_internal, input->ivu_class_id);
+	D_ASSERT(iv_ops != NULL);
+
+	rc = iv_ops->ivo_on_get(ivns_internal, &input->ivu_key, 0,
+				CRT_IV_PERM_WRITE, &iv_value, &user_priv);
+	if (rc != 0) {
+		D_ERROR("ivo_on_get(): "DF_RC"\n", DP_RC(rc));
+		D_GOTO(send_error, rc);
+	}
+
+	d_sgl_buf_copy(&iv_value, &input->ivu_iv_sgl);
+
+	rc = iv_ops->ivo_on_update(ivns_internal, &input->ivu_key, 0, false,
+				   &iv_value, user_priv);
+
+	sync_type = input->ivu_sync_type.iov_buf;
+
+	if (rc == -DER_IVCB_FORWARD ||
+	    (rc == 0 && sync_type->ivs_flags & CRT_IV_SYNC_BIDIRECTIONAL)) {
+
+		D_ALLOC_PTR(update_cb_info);
+		if (update_cb_info == NULL)
+			D_GOTO(put_error, rc = -DER_NOMEM);
+
+		update_cb_info->uci_child_rpc = rpc;
+
+
+		update_cb_info->uci_ivns_internal = ivns_internal;
+		IVNS_ADDREF(ivns_internal);
+
+		update_cb_info->uci_class_id = input->ivu_class_id;
+		update_cb_info->uci_caller_rank = input->ivu_caller_node;
+		update_cb_info->uci_sync_type = *sync_type;
+		update_cb_info->uci_user_priv = user_priv;
+		update_cb_info->uci_iv_value = iv_value;
+		if (rc == -DER_IVCB_FORWARD) {
+			/*
+			 * Forward request to the parent
+			 * Get group version to associate with next_rank.
+			 * Pass it down to crt_ivu_rpc_issue
+			 */
+			D_RWLOCK_RDLOCK(&ivns_internal->cii_grp_priv->gp_rwlock);
+			grp_ver = ivns_internal->cii_grp_priv->gp_membs_ver;
+
+			rc = crt_iv_parent_get(ivns_internal, input->ivu_root_node, &next_rank);
+			D_RWLOCK_UNLOCK(&ivns_internal->cii_grp_priv->gp_rwlock);
+
+			if (rc != 0) {
+				D_DEBUG(DB_TRACE, "crt_iv_parent_get() returned %d\n", rc);
+				IVNS_DECREF(ivns_internal);
```

Comment:
Given you use this twice a ivns_decerf: label would be better here although it sounds like you perhaps don't need the additional reference anyway?

Reply:
ah, sure, though ivns_internal will be an input, so this DECREF will not be needed anymore.